### PR TITLE
[ML] fixing bug when returning multi-doc compressed model definitions

### PR DESCRIPTION
**Pull request by benwtrent**
When deserializing multi-doc compressed model definitions, I periodically receive weird errors regarding bytes references. 

Turns out, when we parse the individual bytes references we parse them into `ByteArrays` which satisfy the `bytes()` method. 

But, `CompositeBytesReference`, when there are multiple BytesReferences, does not allow `bytes()` to be called.

### CODE REVIEW

File: x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/inference/persistence/TrainedModelProvider.java [hunk 1]
```diff
@@ -1222,11 +1223,14 @@ private static QueryBuilder buildQueryIdExpressionQuery(String[] tokens, String
...
         throws ElasticsearchException {
 
-        BytesReference[] bb = new BytesReference[docs.size()];
-        for (int i = 0; i < docs.size(); i++) {
-            bb[i] = docs.get(i).getBinaryData();
-        }
-        BytesReference bytes = CompositeBytesReference.of(bb);
+        // If the user requested the compressed data string, we need access to the underlying bytes.
+        // BytesArray gives us that access.
+        BytesReference bytes = docs.size() == 1
```
Comment (przemekwitek) [inline x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/inference/persistence/TrainedModelProvider.java:1228]: How easy would it be to add a unit test for the multi-doc case?

**Review by przemekwitek — APPROVED** [review]
LGTM

File: x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/inference/persistence/TrainedModelProvider.java (hunk 1 above)
Comment (davidkyle) [reply x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/inference/persistence/TrainedModelProvider.java:1228]: It's a static method so should be simple, can you do this please.

Comment (benwtrent) [reply x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/inference/persistence/TrainedModelProvider.java:1228]: @davidkyle it only occurs when a model definition spreads more than one document. It has to do with the internals of `CompositeBytesReference.of`. `of` returns the passed reference directly (not wrapping it) if the doc number is just 1. 

When there is more than one doc, it actually wraps them, and thus we lose the ability to call: 

```
// We don't store the definition in the same document as the configuration
        if ((params.paramAsBoolean(ToXContentParams.FOR_INTERNAL_STORAGE, false) == false) && definition != null) {
            if (params.paramAsBoolean(DECOMPRESS_DEFINITION, false)) {
                builder.field(DEFINITION.getPreferredName(), definition);
            } else {
                builder.field(COMPRESSED_DEFINITION.getPreferredName(), definition.getBase64CompressedDefinition());
            }
        }
...
private String getBase64CompressedDefinition() throws IOException {
            BytesReference compressedDef = getCompressedDefinition();

            ByteBuffer bb = Base64.getEncoder()
                .encode(ByteBuffer.wrap(compressedDef.array(), compressedDef.arrayOffset(), compressedDef.length()));

            return new String(bb.array(), StandardCharsets.UTF_8);
        }
```

The call to `array()` which returns the bytes, I will fix my PR description. Sorry, last thing of the day and my mind was faster than my fingers typing :D

**Review by davidkyle — APPROVED** [review]
LGTM thanks for adding the tests

**Pull request merged**